Le dataset student_spending contient des informatios sur le 
age	gender	year_in_school	major	monthly_income	financial_aid	tuition	housing	food	transportation	books_supplies	entertainment	personal_care	technology	health_wellness	miscellaneous	preferred_payment_method

In [2]:
import pandas as pd 

df = pd.read_csv("student_spending.csv")
df.head()


Unnamed: 0  age      gender year_in_school             major  \
0           0   19  Non-binary       Freshman        Psychology   
1           1   24      Female         Junior         Economics   
2           2   24  Non-binary         Junior         Economics   
3           3   23      Female         Senior  Computer Science   
4           4   20      Female         Senior  Computer Science   

   monthly_income  financial_aid  tuition  housing  food  transportation  \
0             958            270     5939      709   296             123   
1            1006            875     4908      557   365              85   
2             734            928     3051      666   220             137   
3             617            265     4935      652   289             114   
4             810            522     3887      825   372             168   

   books_supplies  entertainment  personal_care  technology  health_wellness  \
0             188             41             78         134              127   
1             252             74             92         226              129   
2              99            130             23         239              112   
3             223             99             30         163              105   
4             194             48             71          88               71   

   miscellaneous preferred_payment_method  
0             72        Credit/Debit Card  
1             68        Credit/Debit Card  
2            133                     Cash  
3             55       Mobile Payment App  
4            104        Credit/Debit Card

In [38]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 22 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Unnamed: 0                1000 non-null   int64 
 1   age                       1000 non-null   int64 
 2   gender                    1000 non-null   object
 3   year_in_school            1000 non-null   object
 4   major                     1000 non-null   object
 5   monthly_income            1000 non-null   int64 
 6   financial_aid             1000 non-null   int64 
 7   tuition                   1000 non-null   int64 
 8   housing                   1000 non-null   int64 
 9   food                      1000 non-null   int64 
 10  transportation            1000 non-null   int64 
 11  books_supplies            1000 non-null   int64 
 12  entertainment             1000 non-null   int64 
 13  personal_care             1000 non-null   int64 
 14  technology               

In [4]:
print(df.dtypes)

Unnamed: 0                   int64
age                          int64
gender                      object
year_in_school              object
major                       object
monthly_income               int64
financial_aid                int64
tuition                      int64
housing                      int64
food                         int64
transportation               int64
books_supplies               int64
entertainment                int64
personal_care                int64
technology                   int64
health_wellness              int64
miscellaneous                int64
preferred_payment_method    object
dtype: object


In [ ]:
#changer type de colonnez year_in_school
#supprimer les doublons

In [7]:
df.isna().sum()

# Le dataframe contient aucune valeur null 

Unnamed: 0                  0
age                         0
gender                      0
year_in_school              0
major                       0
monthly_income              0
financial_aid               0
tuition                     0
housing                     0
food                        0
transportation              0
books_supplies              0
entertainment               0
personal_care               0
technology                  0
health_wellness             0
miscellaneous               0
preferred_payment_method    0
dtype: int64

In [8]:
df.describe()
#chercher ca veut dire quoi 25%.....

Unnamed: 0          age  monthly_income  financial_aid      tuition  \
count  1000.000000  1000.000000     1000.000000    1000.000000  1000.000000   
mean    499.500000    21.675000     1020.650000     504.771000  4520.395000   
std     288.819436     2.322664      293.841161     287.092575   860.657944   
min       0.000000    18.000000      501.000000       0.000000  3003.000000   
25%     249.750000    20.000000      770.750000     261.000000  3779.750000   
50%     499.500000    22.000000     1021.000000     513.000000  4547.500000   
75%     749.250000    24.000000     1288.250000     751.500000  5285.000000   
max     999.000000    25.000000     1500.000000    1000.000000  6000.000000   

          housing         food  transportation  books_supplies  entertainment  \
count  1000.00000  1000.000000      1000.00000     1000.000000    1000.000000   
mean    696.00600   252.642000       124.63700      174.761000      84.814000   
std     171.21862    86.949606        43.55799       72.404518      37.970451   
min     401.00000   100.000000        50.00000       50.000000      20.000000   
25%     538.75000   175.000000        88.00000      112.000000      54.000000   
50%     704.50000   255.000000       123.00000      175.000000      86.000000   
75%     837.25000   330.000000       162.25000      238.000000     116.000000   
max    1000.00000   400.000000       200.00000      300.000000     150.000000   

       personal_care   technology  health_wellness  miscellaneous  
count    1000.000000  1000.000000      1000.000000    1000.000000  
mean       60.699000   178.304000       114.310000     108.910000  
std        22.898007    71.744441        49.591544      52.412221  
min        20.000000    50.000000        30.000000      20.000000  
25%        41.000000   114.000000        73.000000      63.750000  
50%        62.000000   178.000000       115.000000     110.000000  
75%        80.000000   241.000000       158.000000     153.000000  
max       100.000000   300.000000       200.000000     200.000000

In [ ]:
#dbscan trouver les valeurs abberantes 

Les étudiants ont en moyenne 21 ans et demi (relativement jeunes).
En moyenne, ils reçoivent 1000 dollars de revenu mensuel et 500 dollars supplémentaires d'aide financière.
Leurs frais de scolarité leur coûteront en moyenne 4500 dollars, mais varient considérablement.
Le loyer s'élève en moyenne à 700 dollars, mais varie également autour de 200 dollars. Le loyer le moins cher était de 401 dollars.
Toutes les dépenses semblent avoir des distributions symétriques (la médiane est approximativement égale à la moyenne).
De plus, les quantiles indiquent une distribution uniforme à travers toutes les dépenses.

Les plus gros dépensiers :
Voici une liste des étudiants qui ont le plus dépensé :

In [9]:
df["total"] = df[["tuition", "housing", "food", "transportation", "books_supplies", "entertainment", "personal_care", "technology","health_wellness", "miscellaneous"]].sum(axis=1)

df.sort_values(by="total", ascending=False).head(10)

Unnamed: 0  age      gender year_in_school             major  \
863         863   23        Male         Senior       Engineering   
359         359   21  Non-binary         Junior  Computer Science   
582         582   21  Non-binary         Junior           Biology   
475         475   24      Female       Freshman        Psychology   
881         881   19        Male       Freshman       Engineering   
65           65   19        Male       Freshman           Biology   
170         170   21  Non-binary      Sophomore        Psychology   
471         471   19        Male         Senior  Computer Science   
569         569   21      Female         Senior       Engineering   
723         723   21      Female         Junior           Biology   

     monthly_income  financial_aid  tuition  housing  food  transportation  \
863            1413             80     5976      836   386             194   
359             918            831     5934      949   391              81   
582            1412            671     6000      917   295             191   
475            1331            681     5918      949   335             125   
881             505            293     5861      835   387             181   
65             1092            550     5841      916   150             143   
170             621            456     5893      960   199             124   
471             627            995     5863      937   188             177   
569             940            973     5912      951   352             108   
723             634            330     5961      774   334             193   

     books_supplies  entertainment  personal_care  technology  \
863             266             29             99         213   
359             127            121             64         275   
582             103             22             85         297   
475             258             94             94         171   
881             250             93             91         113   
65              257             90             29         262   
170             278            115             48         193   
471             155            137             97         117   
569              62             62             70         249   
723             105             83             46         166   

     health_wellness  miscellaneous preferred_payment_method  total  
863              109            158                     Cash   8266  
359              141            159        Credit/Debit Card   8242  
582               62            140       Mobile Payment App   8112  
475               34            127       Mobile Payment App   8105  
881               81            171                     Cash   8063  
65               192            177       Mobile Payment App   8057  
170              178             68                     Cash   8056  
471              172            180                     Cash   8023  
569               98            150        Credit/Debit Card   8014  
723              200            150                     Cash   8012

Il me semble qu'il n'y a pas de domaine d'études prédominant pour les gros dépensiers. De plus, les revenus varient fortement. Ce qu'ils ont en commun, ce sont leurs frais de scolarité, chacun étant dans les hauts 5 000 dollars. Regardons donc les dépenses normales.

In [10]:
df["total_spend"] = df["total"] - df["tuition"]
df["total_income"] = df["monthly_income"] + df["financial_aid"]

df.sort_values(by="total_spend", ascending=False).head(10)

Unnamed: 0  age      gender year_in_school             major  \
726         726   18        Male         Senior  Computer Science   
62           62   18      Female         Senior           Biology   
930         930   23      Female         Senior         Economics   
976         976   25        Male         Junior           Biology   
846         846   21  Non-binary       Freshman           Biology   
849         849   19      Female         Junior       Engineering   
888         888   25        Male       Freshman           Biology   
113         113   21      Female       Freshman  Computer Science   
901         901   22        Male         Senior       Engineering   
774         774   25        Male      Sophomore       Engineering   

     monthly_income  financial_aid  tuition  housing  food  ...  \
726             867            646     3014      944   320  ...   
62              733            838     5248      977   298  ...   
930            1123            490     5507      932   370  ...   
976            1418            510     3832      976   370  ...   
846             542            905     3434      987   297  ...   
849            1022            227     4305      982   296  ...   
888            1141            362     3197      926   397  ...   
113             702            273     4977      906   385  ...   
901             894             27     4496      718   373  ...   
774            1257             45     3653      809   373  ...   

     books_supplies  entertainment  personal_care  technology  \
726             232             96             73         298   
62              241             85             78         281   
930             233             91             61         258   
976             218             89             89         283   
846             189            150             26         286   
849             252            149             79         206   
888             150            122             69         258   
113             277            148             41         205   
901             254            139             82         286   
774             246             78             36         275   

     health_wellness  miscellaneous  preferred_payment_method total  \
726              142            183                      Cash  5467   
62               121            168                      Cash  7661   
930              139            187        Mobile Payment App  7919   
976              185             60        Mobile Payment App  6219   
846              185             86                      Cash  5800   
849              152            162                      Cash  6656   
888              173            120        Mobile Payment App  5533   
113              149            116                      Cash  7310   
901              134            198         Credit/Debit Card  6824   
774              158            197         Credit/Debit Card  5972   

     total_spend  total_income  
726         2453          1513  
62          2413          1571  
930         2412          1613  
976         2387          1928  
846         2366          1447  
849         2351          1249  
888         2336          1503  
113         2333           975  
901         2328           921  
774         2319          1302  

[10 rows x 21 columns]

Une chose prédominante semble être le mode de paiement. Je me demande comment les valeurs de paiement sont distribuées dans l'ensemble des données :

In [11]:
df["preferred_payment_method"].value_counts()

preferred_payment_method
Mobile Payment App    350
Credit/Debit Card     340
Cash                  310
Name: count, dtype: int64

il semble que cela soit assez uniformément réparti, avec le mode de paiement le moins préféré étant l'argent liquide. Payer en espèces semble vous faire dépenser plus d'argent. Voyons donc qui perd le plus d'argent chaque mois en fonction des habitudes de dépenses.

In [12]:
df["money_saved"] = df["total_income"] - df["total_spend"]

print(df.groupby("preferred_payment_method").money_saved.mean())

print(df.groupby("preferred_payment_method").money_saved.std())

preferred_payment_method
Cash                 -276.380645
Credit/Debit Card    -291.620588
Mobile Payment App   -242.380000
Name: money_saved, dtype: float64
preferred_payment_method
Cash                  477.635299
Credit/Debit Card     480.101077
Mobile Payment App    485.833766
Name: money_saved, dtype: float64


Il semble que les cartes de crédit ne soient pas bonnes pour votre gestion financière. Les étudiants avec des cartes de crédit dépensent en moyenne 50 dollars de plus par mois que ceux qui utilisent des applications de paiement mobile.

Ensuite, nous voulons connaître la proportion d'étudiants qui ne perdent effectivement pas d'argent chaque mois.

In [13]:
prop_m_saved = len(df.loc[df["money_saved"] > 0, :]) / len(df)

print(f"In total only {prop_m_saved*100}% of students saved money")

In total only 29.5% of students saved money


Maintenant, plongeons plus en profondeur dans les domaines d'études :

In [15]:
df["major"].value_counts()

major
Biology             228
Economics           204
Computer Science    192
Engineering         192
Psychology          184
Name: count, dtype: int64

In [16]:
df["year_in_school"].value_counts()

year_in_school
Senior       254
Freshman     253
Junior       247
Sophomore    246
Name: count, dtype: int64

 La répartition semble suivre une distribution multinomiale avec toutes les probabilités égales. Cela peut ne pas être représentatif, car il y aura plus de premières années que de dernières années. Regardons maintenant la répartition des années d'études après le choix du domaine d'études.

In [17]:
df.groupby("major").year_in_school.value_counts()


major             year_in_school
Biology           Junior            61
                  Freshman          57
                  Sophomore         57
                  Senior            53
Computer Science  Freshman          53
                  Senior            47
                  Junior            46
                  Sophomore         46
Economics         Senior            56
                  Sophomore         56
                  Junior            48
                  Freshman          44
Engineering       Senior            57
                  Junior            48
                  Freshman          47
                  Sophomore         40
Psychology        Freshman          52
                  Sophomore         47
                  Junior            44
                  Senior            41
Name: count, dtype: int64

In [23]:
import plotly.express as px

# Groupement par domaine d'études (major) et année d'études (year_in_school)
resultat = df.groupby("major").year_in_school.value_counts()

# Convertir les résultats en DataFrame pour faciliter la visualisation
resultat_df = resultat.reset_index(name='nombre')

# Créer le sunburst chart
fig = px.sunburst(resultat_df, path=['major', 'year_in_school'], values='nombre',
                  color='nombre', hover_data=['nombre'])

# Afficher le sunburst chart
fig.show()



In [24]:
df.groupby("year_in_school").age.mean()


year_in_school
Freshman     21.490119
Junior       21.712551
Senior       21.590551
Sophomore    21.914634
Name: age, dtype: float64

Oh wow... okay mec, il s'en fout vraiment de la génération de données. Même l'âge n'est pas corrélé à l'année d'études. Je me demande si les étudiants plus âgés gagnent plus.

In [25]:
print("Covariance matrix of age and total income")
print(df[["age", "total_income"]].cov())

print("mean income after age")
print(df.groupby("age").total_income.mean())

Covariance matrix of age and total income
                    age   total_income
age            5.394770     -25.388564
total_income -25.388564  175015.775535
mean income after age
age
18    1525.653226
19    1583.953704
20    1518.351351
21    1527.355932
22    1479.161538
23    1555.515625
24    1529.683824
25    1496.372414
Name: total_income, dtype: float64


D'accord, le revenu total est en fait négativement corrélé au revenu. Normalisons-les pour éviter les problèmes d'échelle. Pour la normalisation, nous allons choisir , chercher sur la normalisation 

In [28]:
df.head()

Unnamed: 0  age      gender year_in_school             major  \
0           0   19  Non-binary       Freshman        Psychology   
1           1   24      Female         Junior         Economics   
2           2   24  Non-binary         Junior         Economics   
3           3   23      Female         Senior  Computer Science   
4           4   20      Female         Senior  Computer Science   

   monthly_income  financial_aid  tuition  housing  food  ...  entertainment  \
0             958            270     5939      709   296  ...             41   
1            1006            875     4908      557   365  ...             74   
2             734            928     3051      666   220  ...            130   
3             617            265     4935      652   289  ...             99   
4             810            522     3887      825   372  ...             48   

   personal_care  technology  health_wellness  miscellaneous  \
0             78         134              127             72   
1             92         226              129             68   
2             23         239              112            133   
3             30         163              105             55   
4             71          88               71            104   

   preferred_payment_method  total total_spend  total_income  money_saved  
0         Credit/Debit Card   7707        1768          1228         -540  
1         Credit/Debit Card   6756        1848          1881           33  
2                      Cash   4810        1759          1662          -97  
3        Mobile Payment App   6665        1730           882         -848  
4         Credit/Debit Card   5828        1941          1332         -609  

[5 rows x 22 columns]

In [32]:
import plotly.express as px

# Remplacez "monthly_income" par le nom de la colonne contenant les données de revenu mensuel dans votre dataframe
fig = px.histogram(df, x="monthly_income", nbins=20, range_x=[0, 2000])

# Afficher l'histogramme
fig.show()



In [33]:
result_dispense_male = df[df["gender"] == "Male"].iloc[:,7:17].mean()
result_dispense_female = df[df["gender"] == "Female"].iloc[:,7:17].mean()
result_dispense_non_binary = df[df["gender"] == "Non-binary"].iloc[:,7:17].mean()

result_dispense_male

tuition            4541.362360
housing             707.755618
food                254.876404
transportation      124.471910
books_supplies      171.165730
entertainment        82.092697
personal_care        60.207865
technology          179.112360
health_wellness     112.837079
miscellaneous       107.030899
dtype: float64

In [34]:
print(type(result_dispense_male))
df_dispense_male = pd.DataFrame(result_dispense_male)
df_dispense_female = pd.DataFrame(result_dispense_female)
df_dispense_non_binary = pd.DataFrame(result_dispense_non_binary)
df_dispense_male = df_dispense_male.rename(columns = {0 :"male"})
df_dispense_female = df_dispense_female.rename(columns = {0 :"female"})
df_dispense_non_binary = df_dispense_non_binary.rename(columns = {0 :"non_binary"})

df_dispense_male

<class 'pandas.core.series.Series'>


male
tuition          4541.362360
housing           707.755618
food              254.876404
transportation    124.471910
books_supplies    171.165730
entertainment      82.092697
personal_care      60.207865
technology        179.112360
health_wellness   112.837079
miscellaneous     107.030899

In [35]:
df_dispense_sexe = pd.merge(df_dispense_male, df_dispense_female, left_index = True, right_index= True)
df_dispense_sexe = pd.merge(df_dispense_sexe, df_dispense_non_binary, left_index = True, right_index= True)
print(df_dispense_sexe)

                        male       female   non_binary
tuition          4541.362360  4494.210526  4523.489097
housing           707.755618   689.504644   689.517134
food              254.876404   247.681115   255.155763
transportation    124.471910   126.040248   123.408100
books_supplies    171.165730   178.987616   174.495327
entertainment      82.092697    85.433437    87.208723
personal_care      60.207865    60.993808    60.947040
technology        179.112360   172.266254   183.482866
health_wellness   112.837079   114.804954   115.445483
miscellaneous     107.030899   108.671827   111.233645


In [36]:
df_dispense_sexe["sexe_who_spend_the_most"] = df_dispense_sexe.idxmax(axis = 1)
df_dispense_sexe

male       female   non_binary sexe_who_spend_the_most
tuition          4541.362360  4494.210526  4523.489097                    male
housing           707.755618   689.504644   689.517134                    male
food              254.876404   247.681115   255.155763              non_binary
transportation    124.471910   126.040248   123.408100                  female
books_supplies    171.165730   178.987616   174.495327                  female
entertainment      82.092697    85.433437    87.208723              non_binary
personal_care      60.207865    60.993808    60.947040                  female
technology        179.112360   172.266254   183.482866              non_binary
health_wellness   112.837079   114.804954   115.445483              non_binary
miscellaneous     107.030899   108.671827   111.233645              non_binary

In [37]:
df_dispense_sexe.sexe_who_spend_the_most.value_counts()


sexe_who_spend_the_most
non_binary    5
female        3
male          2
Name: count, dtype: int64

In [40]:
import plotly.express as px

# Utilisation de Plotly Express pour créer le diagramme en entonnoir à partir du DataFrame df
fig = px.funnel(df, x='gender', y='money_saved')

# Afficher le diagramme en entonnoir
fig.show()


In [41]:
import plotly.express as px

# Utilisation de Plotly Express pour créer le diagramme en entonnoir à partir du DataFrame df
fig = px.funnel(df, x='monthly_income', y='preferred_payment_method')

# Afficher le diagramme en entonnoir
fig.show()


In [42]:
df

Unnamed: 0  age      gender year_in_school             major  \
0             0   19  Non-binary       Freshman        Psychology   
1             1   24      Female         Junior         Economics   
2             2   24  Non-binary         Junior         Economics   
3             3   23      Female         Senior  Computer Science   
4             4   20      Female         Senior  Computer Science   
..          ...  ...         ...            ...               ...   
995         995   22      Female         Senior           Biology   
996         996   19      Female         Senior           Biology   
997         997   20        Male         Junior         Economics   
998         998   22  Non-binary         Senior         Economics   
999         999   24  Non-binary      Sophomore  Computer Science   

     monthly_income  financial_aid  tuition  housing  food  ...  \
0               958            270     5939      709   296  ...   
1              1006            875     4908      557   365  ...   
2               734            928     3051      666   220  ...   
3               617            265     4935      652   289  ...   
4               810            522     3887      825   372  ...   
..              ...            ...      ...      ...   ...  ...   
995            1346            520     3688      969   152  ...   
996            1407            560     3380      508   265  ...   
997             957            393     3497      723   339  ...   
998            1174            612     3649      543   237  ...   
999             541            640     5965      609   270  ...   

     entertainment  personal_care  technology  health_wellness  miscellaneous  \
0               41             78         134              127             72   
1               74             92         226              129             68   
2              130             23         239              112            133   
3               99             30         163              105             55   
4               48             71          88               71            104   
..             ...            ...         ...              ...            ...   
995             42             38         252               65            163   
996             40             98         274               84            135   
997            112             46         284               57             28   
998            129             90         190              101             65   
999            139             43         281               88            145   

     preferred_payment_method  total total_spend  total_income  money_saved  
0           Credit/Debit Card   7707        1768          1228         -540  
1           Credit/Debit Card   6756        1848          1881           33  
2                        Cash   4810        1759          1662          -97  
3          Mobile Payment App   6665        1730           882         -848  
4           Credit/Debit Card   5828        1941          1332         -609  
..                        ...    ...         ...           ...          ...  
995                      Cash   5714        2026          1866         -160  
996        Mobile Payment App   5042        1662          1967          305  
997        Mobile Payment App   5294        1797          1350         -447  
998        Mobile Payment App   5327        1678          1786          108  
999                      Cash   7946        1981          1181         -800  

[1000 rows x 22 columns]